exec/vm.go

```diff
@@ -0,0 +1,332 @@
+	}
+	compiled := vm.compiledFuncs[fnIndex]
+	vm.stack = make([]uint64, compiled.maxDepth)
+	vm.locals = make([]uint64, compiled.totalLocalVars)
+	vm.pc = 0
+	vm.stackTop = -1
+	vm.code = compiled.code
+
+	for i, arg := range args {
+		vm.locals[i] = arg
+	}
+
+	var rtrn interface{}
+	res := vm.execCode(compiled)
+	if compiled.returns {
+		rtrnType := vm.module.GetFunction(int(fnIndex)).Sig.ReturnTypes[0]
+		switch rtrnType {
+		case wasm.ValueTypeI32:
+			rtrn = uint32(res)
+		case wasm.ValueTypeI64:
+			rtrn = uint64(res)
+		case wasm.ValueTypeF32:
+			rtrn = math.Float32frombits(uint32(res))
+		case wasm.ValueTypeF64:
+			rtrn = math.Float64frombits(uint64(res))
+		default:
+			return nil, ErrorInvalidReturnType(int8(rtrnType))
+		}
+	}
+
+	return rtrn, nil
+}
+
+func (vm *VM) execCode(compiled compiledFunction) uint64 {
+	// println("max depth ", compiled.maxDepth)
+outer:
+	for int(vm.pc) < len(vm.code) {
+		// println("stack top is ", vm.stackTop)
+		// print("pc: ", vm.pc, " len is ", len(vm.code), " ")
+		op := vm.code[vm.pc]
+		switch op {
+		case ops.Return:
+			// println("return")
+			break outer
+		case compile.OpJmp:
+			// println("jump")
+			vm.pc = vm.fetchInt64()
+			continue
+		case compile.OpJmpZ:
+			// println("jmpz")
+			target := vm.fetchInt64()
+			if vm.popUint32() == 0 {
+				// println("zero, jumping")
+				vm.pc = target
+				continue
+			}
+		case compile.OpJmpNz:
+			// println("jmpnz")
+			target := vm.fetchInt64()
+			preserveTop := vm.fetchBool()
+			discard := vm.fetchInt64()
+			if vm.popUint32() != 0 {
+				// println("not zero, jumping")
+				vm.pc = target
+				// println("stackTop ", vm.stackTop)
+				var top uint64
+				if preserveTop {
+					top = vm.stack[vm.stackTop]
+				}
+				// println("top: ", top)
+				vm.stackTop -= discard
+				if preserveTop {
+					vm.pushUint64(top)
+				}
+				continue
+			}
+		case ops.BrTable:
+			// println("brtable")
+			index := vm.fetchInt64()
+			label := vm.popInt32()
+			table := vm.compiledFuncs[vm.curFunc].branchTables[index]
+			var target compile.Target
+			if label < int32(len(table.Targets)) {
+				target = table.Targets[int32(label)]
+			} else {
+				target = table.DefaultTarget
+			}
+			vm.pc = target.Addr
+			var top uint64
+			if target.PreserveTop {
+				top = vm.stack[vm.stackTop]
+			}
+			vm.stackTop -= target.Discard
+			if target.PreserveTop {
+				vm.pushUint64(top)
+			}
+			continue
+		case compile.OpDiscard:
+			place := vm.fetchInt64()
+			vm.stackTop -= place
+			// println("discarding ", place, " top is ", vm.stackTop)
+		case compile.OpDiscardPreserveTop:
+			// println("discard preserve top")
+			top := vm.stack[vm.stackTop]
+			place := vm.fetchInt64()
+			vm.stackTop -= place
+			vm.pushUint64(top)
+		default:
+			// f := vm.funcTable[op]
+			// opStr, _ := ops.New(op)
+			// if f == nil {
+			// 	panic(fmt.Sprint("invalid op ", op, opStr.Name))
+			// }
+			// println(opStr.Name)
+			// f()
```

Comment:
Commented out code should be removed if it's no longer needed.

Also, it's common to skip the space after `//` for commented out code. `//` followed by space is reserved for human-facing comments.

Reference:  https://dmitri.shuralyov.com/idiomatic-go#comments-for-humans-always-have-a-single-space-after-the-slashes.